packages/actor-http-cache/lib/ActorHttpCache.ts

```diff
@@ -0,0 +1,205 @@
+    if (await this.has(new Request(action.input, action.init))) {
+      return { time: 1 };
+    }
+    return this.actorHttpFetch.test(action);
+  }
+
+  public async run(action: IActionHttp): Promise<IActorHttpOutput> {
+    // Fetch data
+    return await this.fetchWithCache(action);
+  }
+
+  /**
+   * Fetches a request from the cache or from the web if it is not present in
+   * the cache.
+   * @param request The Request for which you are attempting to find responses
+   * in the Cache. This can be a Request object or a URL.
+   * @param options An object that sets options for the match operation.
+   * @returns A Promise that resolves to the first Response that matches the
+   * request and a boolean indicating if the value was from the cache or not
+   */
+  public async fetchWithCache(
+    action: IActionHttp,
+  ): Promise<Response> {
+    const newRequest = new Request(action.input, action.init);
+    const cacheResult = await this.cacheStorage.get(newRequest);
+    // Nothing is in the cache
+    if (!cacheResult) {
+      const [ body, init ] = await this.fetchDocument(action);
+      try {
+        await this.put(newRequest, body, init);
+      } catch {
+        // Do nothing if it is not storable
+      }
+      return this.newResponse(body, init);
+      // Something is in the cache
+    }
+    // Check if the response is stale
+    const oldPolicy = cacheResult.policy;
+    const oldResponse = this.newResponse(cacheResult.body, cacheResult.init);
+
+    const newRequestWithHashHeaders = requestToRequestWithHashHeaders(
+      new Request(newRequest),
+    );
+    // If the response is stale
+    if (!oldPolicy.satisfiesWithoutRevalidation(newRequestWithHashHeaders)) {
+      // Invaidate the cache if it must be refreshed
+      await this.mediatorHttpInvalidate.mediate({ url: newRequest.url, context: action.context });
+      addHashHeadersToObject(
+        oldPolicy.revalidationHeaders(newRequestWithHashHeaders),
+        newRequest,
+      );
+      // Fetch again with new headers
+      const [ body, init ] = await this.fetchDocument(action);
+      const newResponse = this.newResponse(body, init);
+      const { policy, modified } = oldPolicy.revalidatedPolicy(
+        requestToRequestWithHashHeaders(newRequest),
+        responseToResponseWithHashHeaders(newResponse),
+      );
+      const response = modified ? newResponse : oldResponse;
+      await this.cacheStorage.set(
+        newRequest,
+        { policy, body, init },
+        policy.timeToLive(),
+      );
+      return response;
+    }
+    return oldResponse;
+  }
+
+  /**
+   * Fetches the document and splits it into parts
+   * @param action Fetch action
+   * @returns
+   */
+  private async fetchDocument(action: IActionHttp): Promise<[bodyInit?: BodyInit, responseInit?: ResponseInit]> {
+    const response = await this.actorHttpFetch.run(action);
+    return [
+      await response.text(),
```

Comment:
We will need a more elaborate method for copying response bodies.

Streamability is an important element within Comunica's design.
This means that responses should be parseable in a streaming manner.
One the one hand, this is so that Comunica can already start processing parts of responses from slow servers.
On the other hand, this also means that Comunica can handle infinitely long documents (such as server-sent events).

Therefore, we should copy the body while preserving its streaming nature.
I suspect this one may help here: https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream/tee

Reply:
I tried redoing this using `tee`. Unfortunately, NodeJS bodies don't have a `tee` method and causes the error: `TypeError: response.body?.tee is not a function`

Reply:
Oh, that's annoying, but not surprising actually, given [the workaround we have to add `body.cancel`](https://github.com/comunica/comunica/blob/master/packages/actor-http-fetch/lib/ActorHttpFetch.ts#L100-L110).

We really need to keep streamability though. One low-effort solution would be to monkey-patch body (like `body.cancel`) to add `tee` support, which can be implemented in the usual way of non-blocking cloning Node streams. What do you think?

Reply:
Okay, I think that I've fixed it using NodeJS's streams